rosidl_generator_java/resource/msg.cpp.em

```diff
@@ -339,62 +348,74 @@ jobject @(underscore_separated_type_name)__convert_to_java(@(msg_normalized_type
-  for (size_t i = 0; i < @(member.type.size); ++i) {
+  j@(get_java_name)Array _jarray_@(member.name)_obj = env->New@(get_method_name)Array(@(member.type.size));
+  j@(get_java_name) *_j@(get_java_name)_@(member.name)_buf = (j@(get_java_name) *)malloc(sizeof(j@(get_java_name)) * @(member.type.size));
+  std::copy(_ros_message->@(member.name), _ros_message->@(member.name) + @(member.type.size), _j@(get_java_name)_@(member.name)_buf);
+  env->Set@(get_method_name)ArrayRegion(_jarray_@(member.name)_obj, 0, @(member.type.size), (const j@(get_java_name) *)_j@(get_java_name)_@(member.name)_buf);
+@[      else]@
+  j@(get_java_name)Array _jarray_@(member.name)_obj = env->New@(get_method_name)Array(_ros_message->@(member.name).size);
+  j@(get_java_name) *_j@(get_java_name)_@(member.name)_buf = (j@(get_java_name) *)malloc(sizeof(j@(get_java_name)) * _ros_message->@(member.name).size);
+  std::copy(_ros_message->@(member.name).data, _ros_message->@(member.name).data + _ros_message->@(member.name).size, _j@(get_java_name)_@(member.name)_buf);
+  env->Set@(get_method_name)ArrayRegion(_jarray_@(member.name)_obj, 0, _ros_message->@(member.name).size, (const j@(get_java_name) *)_j@(get_java_name)_@(member.name)_buf);
+@[      end if]@
+  free(_j@(get_java_name)_@(member.name)_buf);
+@[    elif isinstance(member.type.value_type, AbstractGenericString)]@
+@[      if isinstance(member.type, Array)]@
+  jobjectArray _jarray_@(member.name)_obj = (jobjectArray)env->NewObjectArray(@(member.type.size), env->FindClass("java/lang/String"), NULL);
+  for (size_t i = 0; i < @(member.type.size); i++) {
     auto _ros_@(member.name)_element = _ros_message->@(member.name)[i];
 @[      else]@
-  for (size_t i = 0; i < _ros_message->@(member.name).size; ++i) {
+  jobjectArray _jarray_@(member.name)_obj = (jobjectArray)env->NewObjectArray(_ros_message->@(member.name).size, env->FindClass("java/lang/String"), NULL);
+  for (size_t i = 0; i < _ros_message->@(member.name).size; i++) {
     auto _ros_@(member.name)_element = _ros_message->@(member.name).data[i];
 @[      end if]@
-@[      if isinstance(member.type.value_type, AbstractGenericString)]@
-    jobject _jlist_@(member.name)_element = nullptr;
+    jstring _jarray_@(member.name)_element = nullptr;
     if (_ros_@(member.name)_element.data != nullptr) {
-@[        if isinstance(member.type.value_type, AbstractString)]@
-      _jlist_@(member.name)_element = env->NewStringUTF(_ros_@(member.name)_element.data);
-@[        else]@
-      _jlist_@(member.name)_element = env->NewString(_ros_@(member.name)_element.data, _ros_@(member.name)_element.size);
-@[        end if]@
-    }
+@[      if isinstance(member.type.value_type, AbstractString)]@
+      _jarray_@(member.name)_element = env->NewStringUTF(_ros_@(member.name)_element.data);
 @[      else]@
-    jobject _jlist_@(member.name)_element = env->NewObject(
-      _j@(normalized_type)_class_global, _j@(normalized_type)_constructor_global, _ros_@(member.name)_element);
```

Comment:
IIUC, the performance issue is coming from creating new Java objects for each element in the list.